Allow for `quick` parameter on `notify/db/:id` for quick syncing

dpsutton opened the pull request:
fixes #14634.

Allows for a post of `{:table-name "CATEGORIES" :scan schema}` to indicate a quick sync rather than the full sync, analysis, and field-values that `notify/db/:id` normally does.

Also extended so that just posting to `notify/db/<id>` with `{:scan schema}` allows for just the metadata sync rather than a full sync.

dpsutton committed:
Allow for `quick` parameter on `notify/db/:id` for quick syncing

dpsutton committed:
Ensure auth-ed in api tests

salsakran commented:
Overall, this looks like a really good direction.

One thing, I think "quick" is misleading, since if I look at that, when would I not want an api call to be quick. I think maybe something like a `scan` parameter with options of  "schema" or "full", with "full" would keep the api feeling structured. 
Eg,  I can do `/api/notify/1?table_name=orders`,  or `/api/notify/1?table_name=orders&type=full` for a full scan w/ field values
or `/api/notify/1?table_name=orders&scan=schema` would give you a quick sync. only. Open to other parameter names and values but wanted to give a concrete idea.

While we're in here, I think we can future proof it by also providing a sync option for the database as well. This would all end up feeling like a coherent and structured api vs bandaids.

dpsutton committed:
Move to mt/client

dpsutton committed:
Full sync / metadata sync of db

dpsutton committed:
Update tests

dpsutton committed:
errant comment no longer applicable

dpsutton committed:
Rename parameter to {:scan full} or {:scan schema}

dpsutton commented:
Also, is there a proper place to document this? Want to make sure that its discoverable

flamber commented:
@dpsutton Part of the description, then it's automatically added, when the API is updated (manually via a command+PR):
https://github.com/metabase/metabase/blob/master/docs/api-documentation.md

dpsutton committed:
Update api/notify/db docstring and remove extra require

camsaul commented:
> Also, is there a proper place to document this? Want to make sure that its discoverable

I don't know if we have any documentation about the `notify` endpoints -- try searching for that or `MB_API_KEY`. If we do, you can add it there. If not, it might be worth adding a little section about it

camsaul commented on src/metabase/api/notify.clj:
```diff
@@ -4,19 +4,29 @@
+  `Table`. Optional Parameter `:scan` can be `\"full\" or \"schema\" for a full sync or a schema sync, available
+  regardless if a `:table_id` or `:table_name` is passed."
+  [id :as {{:keys [table_id table_name scan]} :body}]
+  (when scan
+    (or (contains? #{"full" :full "schema" :schema} scan)
+        (throw (ex-info "Optional scan parameter must be either \"full\" or \"scan\""
+                        {:status-code 400}))))
```
params are never going to be passed in to the API as keywords, so you don't need to check that.

We have a pattern to check API endpoint parameters using Schema. `defendpoint` will convert a schema map directly after the arglist into a series of appropriate checks. On top of that, it will include those constraints in the API endpoint docstring. Yours should look something like:

```clj
{scan (s/optional (s/enum "full "schema"))}
```

camsaul commented on test/metabase/api/notify_test.clj:
```diff
@@ -1,28 +1,83 @@
 
 (use-fixtures :once (fixtures/initialize :db :web-server))
 
+(defn- do-with-api-key [api-key thunk]
+  (with-redefs [env/env (assoc env/env :mb-api-key api-key)]
+    (thunk)))
+
+(defmacro ^:private with-api-key [api-key & body]
+  `(do-with-api-key ~api-key (fn [] ~@body)))
```
Mentioned in Slack yesterday, but you should use `mt/with-temporary-settings-values` instead

camsaul commented on test/metabase/api/notify_test.clj:
```diff
@@ -1,28 +1,83 @@
+                                      {:headers {"X-METABASE-APIKEY" "testing-api-key"
+                                                 "Content-Type"      "application/json"}}}
+                                     payload)))]
+        (testing "sync just table when table is provided"
+          (let [long-sync-called? (atom false), short-sync-called? (atom false)]
+            (with-redefs [metabase.sync/sync-table!                        (fn [_table] (reset! long-sync-called? true))
+                          metabase.sync.sync-metadata/sync-table-metadata! (fn [_table] (reset! short-sync-called? true))]
+              (post {:scan :full, :table_name table-name})
+              (is @long-sync-called?)
+              (is (not @short-sync-called?)))))
```
Isn't this going to be a race condition? If the sync happens asynchronously (because it's wrapped in `future`) the different sync functions aren't going to get called until some time after the API endpoint completes. I'm surprised you haven't ran in to it locally, but on CircleCI you can count of it randomly not running the other thread for 30 seconds or whatever.

I'm not sure using atoms is going to work here... you could use promises and then `u/deref-with-timeout` them with something generous (like 30 seconds for when Circle decides to be super slow), but you don't want to wait 30 seconds to find out something *didn't* run. Maybe since only one sync function ends up getting called you can use a single promise and have it get set to a value signifying which sync function is ran

dpsutton replied:
oh good point. you're right. I imagine there could be some complex stuff done to submit tasks and put an executor behind a protocol and test that or perhaps make some kind of execute function which takes a thunk and in tests just runs the thunk but in real calls `(thread (thunk))` but fundamentally this is a pretty simple `cond` that we're testing. I'll probably just remove the tests, leaving just the authenticated and 404 tests.

camsaul reviewed (changes requested):
- use `defendpoint` Schema parameter validation instead of rolling your own
- use `mt/with-temporary-settings-values` instead of `with-api-key`
- the tests look prone to race conditions

dpsutton committed:
Use defendpoint schema enforcement instead of bad custom one

dpsutton committed:
Run tests synchronously

dpsutton committed:
Remove custom macro for `mt/with-temporary-setting-values`

dpsutton commented:
@camsaul when you have a chance:
1. uses the validations of schema with defendpoint
```clojure
  {table_id   (s/maybe su/IntGreaterThanZero)
   table_name (s/maybe su/NonBlankString)
   scan       (s/maybe (s/enum "full" "schema"))} ;; s/optional looks like something else
```
2. replaced the custom macro for the api key with `(mt/with-temporary-setting-values [api-key "testing-api-key"]`
3. used the following macro which emits a runtime check whether to run on or off thread
```clojure
(def ^:private ^:dynamic *execute-asynchronously* true)

(defmacro ^:private execute
  "Emit a runtime check whether to run off thread or on thread. Intended to remove race conditions from tests."
  [form]
  `(if *execute-asynchronously* (future ~form) ~form))

(execute (table-sync-fn table))
```
and in testing
```clojure
(binding [notify/*execute-asynchronously* false]
  tests....)
```

dpsutton committed:
Ensure request has non-nil api-key before checking

dpsutton commented:
@camsaul your feedback was handled if you could glance at this when you get time. (know you're busy so thanks for whenever you can get to it)

camsaul commented on src/metabase/api/notify.clj:
```diff
@@ -4,19 +4,37 @@
+
+(def ^:private ^:dynamic *execute-asynchronously* true)
+
+(defmacro ^:private execute
+  "Emit a runtime check whether to run off thread or on thread. Intended to remove race conditions from tests."
+  [form]
+  `(if *execute-asynchronously* (future ~form) ~form))
```
Don't bother changing this, but I have a couple points of feedback on this macro you might want to keep in mind for next time

1. This would be a perfect place to accept an `& body` rather than a single `form`. It costs almost nothing to make it more flexible. Not really necessary here since it's used in only twice, but `& body` is still more idiomatic for macros
2. I've learned to be careful about duplicating forms in macro body, even if we're not worried about multiple evaluation like here because they're in an `if` form. It's good to try to keep the code size down of your macro expansion if you can. Suppose `form` is a hundred lines long -- the macro expansion ends up being 200 lines long. That's inefficient from a code size standpoint, which means it takes longer to compile, takes more space on disk and in memory, and results in more cache misses when ran. Here it makes almost no difference, but I've actually ran into situations where I've got `MethodCodeTooLarge` exceptions during compilation because a macroexpansion generated code that was too big.
3. For one-off tiny macros like this you might want to consider using `clojure.tools.macro/macrolet`
4. Honestly I don't think you really need to make this a macro at all, You could just do something like

   ```clj
   (letfn [(execute [thunk]
             (if *execute-asynchronously*
               (future (thunk))
               (thunk)))]
     ...
     (execute #(table-sync-function table))
     ...)
   ```

dpsutton replied:
great points. as i was writing it i was really wondering if i was justifying not writing the lambda myself. I'm normally pretty stingy on that kind of syntax. Especially when getting it wrongs means it runs synchronously in the current thread during a request. Good points and I appreciate you taking the time.

camsaul commented on src/metabase/server/middleware/auth.clj:
```diff
@@ -40,7 +40,11 @@
-      (handler request respond raise)
-      ;; default response is 403
-      (respond middleware.u/response-forbidden))))
+    (cond (not metabase-api-key)
+          (respond middleware.u/response-forbidden)
```
👍

camsaul reviewed (approved):
Awesome, LGTM

dpsutton committed:
Incorporate feedback from cam about macros

- it is super important that the db/table-syncs are off thread and the
request is not blocked. Getting that macro wrong would be devastating
and it only exists so it can be written `(execute (db-sync-fn
database))` instead of *gasp* `(execute #(db-sync-fn database))`. Used
in three places, and getting it wrong is terrible. Delaying into the
thunk makes clear that this will work just fine.

dpsutton commented:
@camsaul i rewrote it since i realized i was just saving writing `#(sync database)` and i'm normally pretty critical of macros who's only function is to save writing a function literal over a body. thanks for the critical eye. as soon as the tests pass i'll merge.
